### PR TITLE
[Inductor] added aten.geometric_ decomp

### LINKED ISSUE #91671: [RFC] [Inductor] decompose aten.geometric_
### 🚀 The feature, motivation and pitch

Currently `aten.geometric_` fallbacks to `FallbackKernel: torch.ops.aten.geometric_.default`.

```python
import torch

def foo(x, p):
    return x.geometric_(p)
opt_foo = torch.compile(foo)
x = torch.randn(1000)
p = 0.5

print(opt_foo(x, p))
```

`aten.geometric_` is decomposable using `torch.rand` to something like:

```python
def geometric(self, p):
    return torch.floor(torch.log(1 - torch.rand_like(self)) / torch.log(torch.tensor(1 - p))) + 1

print(torch.mean(geometric(torch.randn(100000), p))) # verify E[X] is around 1/p
print(torch.mean(torch.randn(10000).geometric_(p))) # verify E[X] is around 1/p
```

### Alternatives

Please see my comment here https://github.com/pytorch/pytorch/pull/37984#discussion_r1059546209 regarding 

- `log(1 - torch.rand)` for `torch.rand` in `[0, 1)`
- `floor(...) + 1` --vs-- `ceil` 

### Additional context

_No response_

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### PR DESCRIPTION
**Pull request by min-jean-cho**
Fixes #91671

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_decomp/decompositions.py
Comment (min-jean-cho) [reply torch/_decomp/decompositions.py:1967]: Thanks @jgong5, actually `1-U` just transforms `U` in range `[0, 1)` (which is the range of `torch.rand`, `torch.rand_like` ) to `(0, 1]` to avoid `log(0)`.  

`log(1-U)` for `U `in `[0, 1)`
`log(U)` for `U` in `(0, 1]` or `(0, 1)`.

And I think the formula https://pytorch.org/docs/stable/generated/torch.Tensor.geometric_.html here is incorrect.

Comment (jgong5) [reply torch/_decomp/decompositions.py:1967]: Sure, I understood log(U) and log(1-U) part but log(p) vs. log(1-p) part is different. You mean the doc is incorrect. Shall we fix that too?

Comment (min-jean-cho) [reply torch/_decomp/decompositions.py:1967]: Do you mean `log(q) = log(1-p)`? `q = 1 - p`

Yes, I think the doc is incorrect (pretty sure..) https://github.com/pytorch/pytorch/pull/92049

Comment (min-jean-cho) [reply torch/_decomp/decompositions.py:1967]: Oh, you meant `log(p)` if the doc were correct

Comment (lezcano) [reply torch/_decomp/decompositions.py:1967]: The formula in the docs is certainly non-standard. See https://en.wikipedia.org/wiki/Geometric_distribution. We should improve the docs with a description of what does this model (in this case, probability until the first miss).

For the formula I'd suggest two things. First: How is it implemented in core? And second, consider using `torch.log1p` for better stability around 0.

Comment (min-jean-cho) [reply torch/_decomp/decompositions.py:1967]: > The formula in the docs is certainly non-standard. See https://en.wikipedia.org/wiki/Geometric_distribution. We should improve the docs with a description of what does this model (in this case, probability until the first miss). How is it implemented in core?

The [core ](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/TransformationHelper.h#L149-L153)implements a standard geometric distribution, the first one described in the linked wiki.  
`P{X=i} = p * q^(i-1), i >= 1, where q = 1 -p` 
`p: probability of success`
`q: probability of failure`
`P{X=i}: probability of first success on the ith trial`

Note that, the inplace geometric https://pytorch.org/docs/stable/generated/torch.Tensor.geometric_.html models the first one described in the linked wiki while `torch.distributions.geometric.Geometric` https://pytorch.org/docs/stable/distributions.html?highlight=geometric#torch.distributions.geometric.Geometric models the second one described in the linked wiki. 

>  And second, consider using torch.log1p for better stability around 0

Thanks @lezcano !

Comment (min-jean-cho) [reply torch/_decomp/decompositions.py:1967]: Sorry, had a closer look at what `torch.log1p` does. And I don't this `torch.log1p` is suitable to just add `1+input`.

i.e., `floor( log(2-U) / log(2-p) ) + 1` is certainly not the same as  `floor( log(1-U) / log(1-p) ) + 1`

Comment (lezcano) [reply torch/_decomp/decompositions.py:1967]: See how it's implemented here:
https://pytorch.org/docs/stable/_modules/torch/distributions/geometric.html#Geometric

Comment (min-jean-cho) [reply torch/_decomp/decompositions.py]: Thanks @lezcano, done, makes sense.

File: torch/_decomp/decompositions.py
Comment (lezcano) [inline torch/_decomp/decompositions.py:1913]: Actually, if `p` is a number, it's better to do `math.log1p(-p)`.
Also, as in the other function, it may be better to implement the `in-place` version in terms of the out-of-place function.

**Review by jgong5 — APPROVED** [review]
LGTM as long as you address the remaining comments from @lezcano

File: torch/_refs/__init__.py
Comment (min-jean-cho) [inline torch/_refs/__init__.py:5124]: Actually, I realize `torch.finfo` won't do for integer dtype. So it's probably better to represent `log(1-U)` as `log1p(-U)` similar to as in exponential.

**Review by jansel — APPROVED** [review]

File: torch/_refs/__init__.py
Comment (ngimel) [inline torch/_refs/__init__.py:5125]: this is wrong, `aten.geometric` is a non-inplace op, the inplace variant will be added with `_make_inplace`, so there should be no `copy_` here. 
@lezcano should there be a test to flag this kind of wrong behavior?

File: torch/_inductor/decomposition.py
Comment (ngimel) [inline torch/_inductor/decomposition.py:526]: should we also add `aten.geometric`, after functionalization the calls to inplace op should be replaced with outplace variant

**Review by ngimel — CHANGES_REQUESTED** [review]

File: torch/_refs/__init__.py
Comment (lezcano) [reply torch/_refs/__init__.py:5125]: It'd be great to have such a test. Let me see if I can put something of that sort next week. This should be a good complement for https://github.com/pytorch/pytorch/pull/92003

Comment (lezcano) [reply torch/_refs/__init__.py:5125]: Addressed in https://github.com/pytorch/pytorch/pull/92243

Comment (min-jean-cho) [reply torch/_refs/__init__.py]: Oops, thanks for the catch, done.

File: torch/_inductor/decomposition.py
Comment (min-jean-cho) [reply torch/_inductor/decomposition.py]: Done, we should also do the same for other in-place random ops

**Review by lezcano — APPROVED** [review]
Nice!

**Comment by lezcano** [comment]
Actually, could you also add the relevant OpInfo for this ref (and the operation itself if it doesn't have one)?

**Review by jansel — APPROVED** [review]

**Review by lezcano — APPROVED** [review]

**Comment by min-jean-cho** [comment]
[Geometric] Hi all, thank you for the review. I see all tests have passed now - I've made few minor changes since the last approvals, could you help have a final scan ?

**Comment by min-jean-cho** [comment]
I will go ahead and merge this as the minor changes after the last approvals are necessary (same changes as #92047)

**Pull request closed without merging**
